drivers/gpio/gpio_mcp230xx.c

```diff
@@ -0,0 +1,377 @@
+		(struct mcp230xx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	buf = drv_data->reg_cache.gpio;
+	buf = (buf & ~mask) | (mask & value);
+
+	ret = write_port_regs(dev, REG_GPIO, buf);
+	if (ret == 0) {
+		drv_data->reg_cache.gpio = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+	return ret;
+}
+
+static int mcp230xx_port_set_bits_raw(const struct device *dev, uint32_t mask)
+{
+	return mcp230xx_port_set_masked_raw(dev, mask, mask);
+}
+
+static int mcp230xx_port_clear_bits_raw(const struct device *dev,
+					uint32_t mask)
+{
+	return -mcp230xx_port_set_masked_raw(dev, mask, 0);
+}
+
+static int mcp230xx_port_toggle_bits(const struct device *dev, uint32_t mask)
+{
+	struct mcp230xx_drv_data *drv_data =
+		(struct mcp230xx_drv_data *)dev->data;
+	uint16_t buf;
+	int ret;
+
+	if (k_is_in_isr()) {
+		return -EWOULDBLOCK;
+	}
+
+	k_sem_take(&drv_data->lock, K_FOREVER);
+
+	buf = drv_data->reg_cache.gpio;
+	buf ^= mask;
+
+	ret = write_port_regs(dev, REG_GPIO, buf);
+	if (ret == 0) {
+		drv_data->reg_cache.gpio = buf;
+	}
+
+	k_sem_give(&drv_data->lock);
+
+	return ret;
+}
+
+static int mcp230xx_pin_interrupt_configure(const struct device *dev,
+					    gpio_pin_t pin,
+					    enum gpio_int_mode mode,
+					    enum gpio_int_trig trig)
+{
+	return -ENOTSUP;
+}
+
+static const struct gpio_driver_api api_table = {
+	.pin_configure = mcp230xx_pin_cfg,
+	.port_get_raw = mcp230xx_port_get_raw,
+	.port_set_masked_raw = mcp230xx_port_set_masked_raw,
+	.port_set_bits_raw = mcp230xx_port_set_bits_raw,
+	.port_clear_bits_raw = mcp230xx_port_clear_bits_raw,
+	.port_toggle_bits = mcp230xx_port_toggle_bits,
+	.pin_interrupt_configure = mcp230xx_pin_interrupt_configure,
+};
+
+/**
+ * @brief Initialization function of MCP230XX
+ *
+ * @param dev Device struct.
+ * @return 0 if successful. Otherwise <0 is returned.
+ */
+static int mcp230xx_init(const struct device *dev)
+{
+	const struct mcp230xx_config *config = dev->config;
+	struct mcp230xx_drv_data *drv_data =
+		(struct mcp230xx_drv_data *)dev->data;
+
+	if (config->ngpios != 8 && config->ngpios != 16) {
+		LOG_ERR("Invalid value ngpios=%u. Expected 8 or 16!", config->ngpios);
+		return -EINVAL;
+	}
+
+	drv_data->i2c = device_get_binding((char *)config->i2c_dev_name);
+	if (!drv_data->i2c) {
+		LOG_ERR("Unable to get SPI device");
+		return -ENODEV;
+	}
+
+	drv_data->i2c_cfg = I2C_SPEED_SET(I2C_SPEED_STANDARD) | I2C_MODE_MASTER;
+
+	if (i2c_configure(drv_data->i2c, drv_data->i2c_cfg)) {
+		LOG_ERR("I2C configure failed");
+		return -ENODEV;
+	}
+
+	k_sem_init(&drv_data->lock, 1, 1);
+
+	return 0;
+}
+
+#define MCP230XX_INIT(inst)					       \
+	static struct mcp230xx_config mcp230xx_##inst##_config = {     \
+		.config = {					       \
+			.port_pin_mask =			       \
+				GPIO_PORT_PIN_MASK_FROM_DT_INST(inst), \
+		},						       \
+		.i2c_dev_name = DT_INST_BUS_LABEL(inst),	       \
```

Comment:
In the new device drivers we should use
```
.i2c = DEVICE_DT_GET(DT_INST_BUS(idx)),			   \
```
`device_get_binding` in `mcp230xx_init` will no longer be necessary. Instead in the init function we should add
```
	if (!device_is_ready(drv_cfg->i2c)) {
		LOG_ERR("%s is not ready", drv_cfg->i2c->name);
		return -ENODEV;
	}

```